interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	return prompt, nil
+}
+
+// Reply resolves the prompt with the given ID using the given outcome by
+// sending a reply to all associated listener requests, then removing the
+// prompt from the prompt DB.
+//
+// Records a notice for the prompt, and returns the prompt's former contents.
+func (pdb *PromptDB) Reply(user uint32, id string, outcome prompting.OutcomeType) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	allow, err := outcome.IsAllow()
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
+	pdb.notifyPrompt(user, id)
+	return prompt, nil
+}
+
+var sendReply = func(listenerReq *listener.Request, reply interface{}) error {
+	return listenerReq.Reply(reply)
+}
+
+// HandleNewRule checks if any existing prompts are satisfied by the given rule
+// contents and, if so, sends back a decision to their listener requests.
+//
+// A prompt is satisfied by the given rule contents if the user, snap,
+// interface, and path of the prompt match those of the rule, and if either the
+// outcome is "allow" and all of the prompt's permissions are matched by those
+// of the rule contents, or if the outcome is "deny" and any of the permissions
+// match.
+//
+// Records a notice for any prompt which was satisfied, or which had some of
+// its permissions satisfied by the rule contents. In the future, only the
+// remaining unsatisfied permissions of a partially-satisfied prompt must be
+// satisfied for the prompt as a whole to be satisfied.
+//
+// Returns the IDs of any prompts which were fully satisfied by the given rule
+// contents.
+func (pdb *PromptDB) HandleNewRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType) ([]string, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	allow, err := outcome.IsAllow()
+	if err != nil {
+		return nil, err
+	}
+	satisfiedPromptIDs := []string{}
+	userEntry, ok := pdb.perUser[user]
+	if !ok {
+		return satisfiedPromptIDs, nil
+	}
+	for id, prompt := range userEntry.ByID {
+		if !(prompt.Snap == snap && prompt.Interface == iface) {
+			continue
+		}
+		matched, err := constraints.Match(prompt.Constraints.Path)
+		if err != nil {
+			return nil, err
+		}
+		if !matched {
+			continue
+		}
+		modified := prompt.Constraints.subtractPermissions(constraints.Permissions)
+		if !modified {
+			continue
+		}
+		pdb.notifyPrompt(user, id)
```

Comment:
why does this notify first and reply after while Reply does the reverse? if that wasn't the case maybe the two could also share some code ?

Reply:
I think there were two reasons for the differences:

1. Here, we `continue` if some but not all of the permissions of the prompt are satisfied by the new rule, in which case we want to tell the clients to re-check the prompt details via the prompting API. We need to record the notice before the `continue`, so it can't be any closer to the `sendReply` call.
2. In `Reply`, we're sending a user-given reply to the listener. In the off-chance that this results in an error (could be possible in the future if we pass more than a `bool` as the reply), we don't actually want to resolve the prompt, since the listener request itself has not yet been correctly resolved. So we return an error before removing the prompt of recording a notice. But we shouldn't encounter errors when replying based on a new rule, since we're theoretically in control of the reply format more. But really we should be fully vetting user replies before sending anything to the listener, so this is potentially a moot point.

Given these considerations, I think the best we could do is move the `notifyPrompt` call to just before the listener req reply loop in `Reply`, and assume there will be no errors. Or leave the errors return in place in the off-chance that it occurs -- there's no harm in re-recording a notice, telling the client to check again (in this case nothing will have changed, though maybe the client will send a better reply next time).

I'm a bit torn, what do you think about this?

Note to self: in the error case in (2), the listener request can't actually be replied to again, even though the first reply failed, so need to fix that in the listener. Possibilities are either sending an auto-deny on error or using a mutex instead of `atomic.CompareAndSwap`.

Reply:
Fundamentally, it doesn't matter when within the function the notice is recorded, since the lock is held any future API requests to get more info will block until we're done with this function anyway.